python/social_bot/tasks.py

```diff
@@ -499,18 +524,59 @@ def task_specific_observation(self, agent):
         Args:
             agent (GazeboAgent): the agent
         Returns:
-            np.array of the extra observations will be added into the
-            observation besides self states, for the non-image case
+            np.array, the observations of the task for non-image case
         """
         goal = self._world.get_model(self._goal_name)
-        pose = np.array(goal.get_pose()[0]).flatten()
+        goal_first = not agent._with_language
+        if goal_first:  # put goal first
+            pose = np.array(goal.get_pose()[0]).flatten()
+        else:  # has language input, don't put goal first
+            pose = None
 
         for name in self._additional_observation_list:
+            if goal_first and name == self._goal_name:
+                continue
             obj = self._world.get_model(name)
             obj_pos = np.array(obj.get_pose()[0]).flatten()
-            pose = np.concatenate((pose, obj_pos), axis=0)
-
-        return pose
+            if pose is None:
+                pose = obj_pos
+            else:
+                pose = np.concatenate((pose, obj_pos), axis=0)
+
+        agent_pose = np.array(agent.get_pose()).flatten()
+        if self._use_egocentric_states:
+            yaw = agent_pose[5]
+            # adds egocentric velocity input
+            vx, vy, vz, a1, a2, a3 = np.array(agent.get_velocities()).flatten()
+            rvx, rvy = agent.get_egocentric_cord_2d(vx, vy, -yaw)
+            obs = [rvx, rvy, vz, a1, a2, a3]
+            # adds objects' (goal's as well as distractions') egocentric
+            # coordinates to observation
+            while len(pose) > 1:
+                x = pose[0] - agent_pose[0]
+                y = pose[1] - agent_pose[1]
+                rotated_x, rotated_y = agent.get_egocentric_cord_2d(x, y, -yaw)
+                if self._egocentric_perception_range > 0:
+                    dist = math.sqrt(rotated_x * rotated_x + rotated_y * rotated_y)
+                    rotated_x /= dist
+                    rotated_y /= dist
+                    magnitude = 1. / dist
+                    if rotated_x < np.cos(
+                        self._egocentric_perception_range / 180. * np.pi):
+                        rotated_x = 0.
+                        rotated_y = 0.
+                        magnitude = 0.
+                    obs.extend([rotated_x, rotated_y, magnitude])
+                else:
+                    obs.extend([rotated_x, rotated_y])
+                pose = pose[3:]
+            obs = np.array(obs)
+        else:
+            agent_vel = np.array(agent.get_velocities()).flatten()
+            joints_states = agent.get_internal_states()
+            obs = np.concatenate(
+                (pose, agent_pose, agent_vel, joints_states), axis=0)
```

Comment:
If use_egocentric_states, joints_states (the sin and cos of wheel joint angles and wheel speed) is not included. Is this on purpose?

Reply:
Yes, I just haven't experimented it.  But good point.  Let me try and see if that improves things!